app/uk/gov/hmrc/gform/testonly/DebugController.scala

```diff
@@ -75,4 +78,217 @@ class DebugController(
+        }
+
+        val formComponentMetadatas: List[ExprMetadata] =
+          formModel.allFormComponents.flatMap(AllFormComponentExpressions.unapply).flatten
+
+        val exprMetadatas: List[ExprMetadata] = formModel.exprsMetadata ++ formComponentMetadatas
+
+        val inspectedExprs: List[InspectedFirstExpr] =
+          exprMetadatas.map(_.expr).map(inspectExpr(_, formModel))
+
+        val includeIfs: List[IncludeIf] = formModel.allIncludeIfsWithDependingFormComponents.map(_._1)
+
+        val inspectedIncludeIfs: List[BooleanExprInfo] =
+          includeIfs.map(includeIf => inspectBooleanExpr(includeIf.booleanExpr, formModel))
+
+        val validIfs: List[ValidIf] = formModel.allValidIfs.flatMap(_._1)
+
+        val inspectedValidIfs: List[BooleanExprInfo] =
+          validIfs.map(validIf => inspectBooleanExpr(validIf.booleanExpr, formModel))
+
+        val page =
+          html.debug.verify.verify(inspectedExplicitExprs, inspectedExprs, inspectedIncludeIfs, inspectedValidIfs)
+        Future.successful(Ok(page))
+
+    }
+
+  def inspectBooleanExpr(booleanExpr: BooleanExpr, formModel: FormModel[DataExpanded]): BooleanExprInfo = {
+
+    def loop(booleanExpr: BooleanExpr): List[InspectedBooleanExpr] = {
+      def inner(expr1: Expr, expr2: Expr): InspectedBooleanExpr.Binary =
+        InspectedBooleanExpr.Binary(booleanExpr, inspectExpr(expr1, formModel), inspectExpr(expr2, formModel))
+      booleanExpr match {
+        case Equals(left, right)              => inner(left, right) :: Nil
+        case GreaterThan(left, right)         => inner(left, right) :: Nil
+        case GreaterThanOrEquals(left, right) => inner(left, right) :: Nil
+        case LessThan(left, right)            => inner(left, right) :: Nil
+        case LessThanOrEquals(left, right)    => inner(left, right) :: Nil
+        case Not(e)                           => loop(e)
+        case Or(left, right)                  => loop(left) ::: loop(right)
+        case And(left, right)                 => loop(left) ::: loop(right)
+        case IsTrue                           => Nil
+        case IsFalse                          => Nil
+        case Contains(multiValueField, value) =>
+          InspectedBooleanExpr
+            .Contains(booleanExpr, inspectExpr(multiValueField, formModel), inspectExpr(value, formModel)) :: Nil
+        case In(formCtx, dataSource) => Nil
+      }
+    }
+
+    BooleanExprInfo(booleanExpr, loop(booleanExpr))
+  }
+
+  def inspectExpr(expr: Expr, formModel: FormModel[DataExpanded]): InspectedFirstExpr = {
+    val firstOperandInfo: TypeInfo = formModel.toFirstOperandTypeInfo(expr)
+
+    def inner(expr1: Expr, expr2: Expr) = {
+      val typeInfo1 = formModel.toFirstOperandTypeInfo(expr1)
+      val typeInfo2 = formModel.toFirstOperandTypeInfo(expr2)
+      List(typeInfo1, typeInfo2) ::: loop(expr1) ::: loop(expr2)
+    }
+
+    def loop(expr: Expr): List[TypeInfo] =
```

Comment:
this is not a loop, could be renamed to avoid confusion

Reply:
Done